### DISCUSSION TITLE
[TPAC Prep] CSS modules

### DISCUSSION THREAD

**Discussion by o-t-w** [w3c/webcomponents-cg#70] [General]
This is a summary of the current status of CSS modules

CSS Module Scripts allow JavaScript modules to import style sheets. They can then be applied to a document or shadow root using adoptedStyleSheets in the same way as constructable style sheets.

```js
import styles from "./styles.css" with { type: "css" };
document.adoptedStyleSheets = [styles];
```

## Motivation
- Many developers want to write their CSS in separate `.css` files, rather than as strings in JavaScript. That is possible today using a `<link>` tag, but that approach is [not recommended](https://lit.dev/docs/components/styles/#external-stylesheet).
- CSS modules will allow styles to be loaded in component definitions without adding `<style>` or `<link>` elements that need to be created in each element instance.
- Without CSS-in-JS it's difficult to ensure that CSS is loaded before the component is defined and rendered. CSS module scripts provide the same ordering guarantee.
- Component authors may need to import a resource that is only available as an external .css file, and don't have control over the resource to wrap it in a JS module.

For further motivational details, see this explainer document: [webcomponents/css-modules-v1-explainer.md at gh-pages · WICG/webcomponents](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/css-modules-v1-explainer.md#why-are-css-modules-needed).

I've recently written about the import attributes [here](https://fullystacked.net/posts/import-attributes/).

## Open issues
There is one open issue: [define behavior of @import in CSS modules 
](https://github.com/WICG/webcomponents/issues/870)

There is an open pull request for [updating the HTML spec](https://github.com/whatwg/html/pull/9486) which would resolve the [remaining issue](https://github.com/whatwg/html/issues/7233) about import attributes. 
Seeing as CSS modules are unlikely to land until JSON modules land, these two open PRs are also relevant:
- https://github.com/whatwg/fetch/pull/1691
- https://github.com/w3c/webappsec-csp/pull/611

## Browser support
CSS modules have two prerequisites: constructable stylesheets and import attributes. 

Constructable stylesheets are supported in all browsers. 

An older syntax of import attributes, then known as import assertions, shipped in Chrome 91. An update to the syntax is currently implemented in Chrome Canary behind a flag. Support for the older syntax is not being removed from Chrome. 

Chrome 93 added support for CSS modules.

Firefox are currently [working on support for import attributes](https://bugzilla.mozilla.org/show_bug.cgi?id=1844128), but will start with JSON modules and [ship CSS modules](https://bugzilla.mozilla.org/show_bug.cgi?id=1720570) at a later date. 

[WebKit position](https://github.com/WebKit/standards-positions/issues/77) 
A PR has been merged adding support for [import attributes](https://github.com/WebKit/WebKit/pull/16730).
No developers are currently assigned to the [WebKit implementation](https://bugs.webkit.org/show_bug.cgi?id=227967) of CSS modules.  

## [Multiple stylesheets per file](https://github.com/w3c/csswg-drafts/issues/5629)
Currently, when using CSS module scripts in Chrome, every CSS import makes a separate HTTP request. Allowing multiple stylesheets within a single CSS file would allow bundlers to merge all the CSS into a single file. The stylesheets would then be available as named exports.  
```js
import {sheet1, sheet2} from './styles1and2.css' with {type: 'css'};
```
In April the CSS working group discussed this proposal and resolved "Accept @sheet with URL fragment referencing rule. Exact details to be in the Cascade spec". It will be part of the CSS Cascade Level 6 specification but has not been added yet. There are no open issues. 

It would be worth fleshing out the use-case and benefit for this feature outside of web components. Given that PostCSS can already concatonate `@import` statements into a single CSS file, and use of multiple stylesheets per file would still make use of a build tool to create the single file, the benefit isn't too clear.   

## Related specs and proposals
- [Declarative CSS module scripts](https://github.com/WICG/webcomponents/issues/939)

**Comment by Westbrook** [comment]
Great intro to this as a spec. Would you be able to share some specifics around the use case this unlocks and why it is valuable?

An overview of the status/blockers on https://github.com/whatwg/html/issues/7233 would be helpful in outlining for implementors the work we’d ask for them here. 

Also, I’m a big fan of the capabilities outlined by https://github.com/w3c/csswg-drafts/issues/5629. Do you have any thought on that being a part of the work needed to call this “done”?

**Comment by o-t-w** [comment]
I don't think the rationale or use case has changed since the previous report, so I've basically copied that over.
I've added a bit of info about the current status of https://github.com/whatwg/html/issues/7233 and https://github.com/w3c/csswg-drafts/issues/5629. I don't think anything about https://github.com/whatwg/html/issues/7233 is actually contentious, the PR just needs the addition of tests as far as I can tell.
